File: experimental/hashly.js
var xxhash = require('xxhash').hash;

var BUCKET_SIZE = 32;
var SEED = 12345;
var BIT = 32;

var hashF = function(key) {
  return xxhash(new Buffer('' + key), SEED);
};

var Item = function(key, val, hash) {
  this.key = key;
  this.val = val;
  this.hash = hash;
};

var Bucket = function(defaultValue) { // members are Item
  this.defaultValue = defaultValue;
  Array.call(this, BUCKET_SIZE);
};
Bucket.prototype = Object.create(Array.prototype);

Bucket.prototype.append = function(newItem, count) {
  var newHash = newItem.hash;
  for (var i = count; i--;) {
    var item = this[i];
    if (item.hash === newHash) { // or compare the key
      this[i] = newItem;
      return false;
    }
  }
  this[count] = newItem;
  return true;
};

Bucket.prototype.find = function(hash, count) {
  for (var i = count; i--;) {
    var item = this[i];
    if (item.hash === hash) { // or compare the key
      return item.val;
    }
  }
  return this.defaultValue;
};

Bucket.prototype.remove = function(hash, count) {
  for (var i = count; i--;) {
    var item = this[i];
    if (item.hash === hash) { // or compare the key
      this[i] = this[count - 1];
      return true;
    }
  }
  return false;
};

var Node = function(defaultValue, oldPtr) {
  this.count = 0; // maybe we can have another counter to fill items from both top and bottom of the bucket. In order to reduce the number of linear probing, check the next highest bit and decide whether we should put this item from the bottom or top
  this.ptr = oldPtr || new Bucket(defaultValue);
};

var Hashly = module.exports = function(defaultValue) {
  this.arrayedTree = [new Node(defaultValue)]; // http://en.wikipedia.org/wiki/Binary_tree
  this.minHeight = 0; // real min height - 1
  //this.minHeightRequest = 0;
  this.defaultValue = defaultValue;
};

Hashly.prototype._updateMinHeight = function(decrement) {
  //if (!decrement) {
  //  if ((1 << this.minHeight) > (++this.minHeightRequest)) {
  //    return;
  //  }
  //  this.minHeightRequest = 0;
  //}
  this.minHeight -= decrement;
  var minHeight = this.minHeight;
  var upper = (1 << (minHeight + 1)) - 1;
  var lower = (1 << minHeight) - 1;
  var arrayedTree = this.arrayedTree;
  var res = true;
  for (var i = lower; i < upper && res; i++) {
    res &= !(arrayedTree[i].ptr);
  }
  this.minHeight += res;
};

Hashly.prototype.get = function(key) {
  var hash = hashF(key);
  var minHeight = this.minHeight;
  var i = (1 << minHeight) - 1 + (minHeight && (hash >>> (BIT - minHeight)));
  for (var bit = minHeight; bit < BIT; bit++) {
    var arrayedTree = this.arrayedTree;
    var node = arrayedTree[i];
    var ptr = node.ptr;
    if (ptr) {
      return ptr.find(hash, node.count);
    }
    var mask = 0x80000000 >>> bit;
    var right = !! (hash & mask);
    i = (i << 1) + 1 + right; // child node
  }
};

Hashly.prototype.set = function(key, val) {
  var hash = hashF(key);
  var minHeight = this.minHeight;
  var i = (1 << minHeight) - 1 + (minHeight && (hash >>> (BIT - minHeight)));
  for (var bit = minHeight; bit < BIT; bit++) {
    var arrayedTree = this.arrayedTree;
    var node = arrayedTree[i];
    var ptr = node.ptr;
    var mask = 0x80000000 >>> bit;
    if (ptr) {
      var count = node.count;
      if (count < BUCKET_SIZE) {
        node.count += ptr.append(new Item(key, val, hash), count); // increment if append returns true
        return true;
      } else { // it is full, so add a bucket
        node.count = 0;
        node.ptr = null;
        var defaultValue = this.defaultValue;
        var i_shift = i << 1;
        var leftChild = arrayedTree[i_shift | 1] = new Node(defaultValue, ptr);
        var rightChild = arrayedTree[i_shift + 2] = new Node(defaultValue);
        var rPtr = rightChild.ptr;
        var rCount = rightChild.count;
        for (var j = count; j--;) {
          var item = ptr[j];
          if (item.hash & mask) {
            rPtr.append(item, rCount++);
            ptr[j] = ptr[--count];
          }
        }
        leftChild.count = count;
        rightChild.count = rCount;
        this._updateMinHeight(false);
      }
    }
    var right = !! (hash & mask);
    i = (i << 1) + 1 + right; // child node
  }
};

Hashly.prototype.has = function(key) {
  return this.get(key) !== this.defaultValue;
};

Hashly.prototype.del = function(key) {
  var hash = hashF(key);
  var minHeight = this.minHeight;
  var i = (1 << minHeight) - 1 + (minHeight && (hash >>> (BIT - minHeight)));
  for (var bit = minHeight; bit < BIT; bit++) {
    var arrayedTree = this.arrayedTree;
    var node = arrayedTree[i];
    var ptr = node.ptr;
    if (ptr) {
      var res = ptr.remove(hash, node.count);
      if (res && ! (--node.count) && i) { // if it becomes empty after deletion
        var left = i & 1; // left is always an odd number
        var i_1 = i - 1;
        var sibling = arrayedTree[i_1 + left * 2];
        var parent = i_1 >>> 1;
        if (sibling.ptr) { // if sibling does not have children
          arrayedTree[parent] = sibling;
          this._updateMinHeight(true);
        }
      }
      return res;
    }
    var mask = 0x80000000 >>> bit;
    var right = !! (hash & mask);
    i = (i << 1) + 1 + right; // child node
  }
};

File: experimental/xcode/hashly/hashly/hashly.cpp
//
//  hashly.cpp
//  hashly
//
//  Created by Kazuyuki Tanimura on 7/4/14.
//  Copyright (c) 2014 Kazuyuki Tanimura. All rights reserved.
//

#include "hashly.h"
#include "xxhash.h"

Bucket::Bucket(double defaultValue) : _defaultValue(defaultValue) {
  items = new Item[BUCKET_SIZE];
  count = 0;
}
Bucket::~Bucket() {
  delete[] items;
}

double Bucket::find(uint32_t hash) {
  for (uint8_t i = count; i--;) {
    if (items[i].hash == hash) { // or compare the key
      return items[i].val;
    }
  }
  return _defaultValue;
}

bool Bucket::insert(Item* newItem) {
  items[count].key = newItem->key;
  items[count].val = newItem->val;
  items[count].hash = newItem->hash;
  count++;
  return true;
}

bool Bucket::insert(const char *key, double val, uint32_t hash) {
  for (uint8_t i = count; i--;) {
    if (items[i].hash == hash) { // or compare the key
      items[i].val = val;
      return false;
    }
  }
  /**
   * We do not do this in a copy constructor in order to avoid allocating
   * extra memory for the right hand side values (i.e., key, val, hash)
   */
  items[count].key = key;
  items[count].val = val;
  items[count].hash = hash;
  count++;
  return true;
}

bool Bucket::remove(uint32_t hash) {
  for (uint8_t i = count; i--;) {
    if (items[i].hash == hash) { // or compare the key
      items[i] = items[count - 1];
      count--;
      return true;
    }
  }
  return false;
}

Hashly::Hashly(double defaultValue) : _defaultValue(defaultValue) {
  /**
   * Do not use the new operator for arrayedTree (i.e., new Bucket*[size])
   * since we do not want to call the default constructor of Bucket.
   * Maybe we just should use std::vector here, but we want to make sure
   * that we can exploit the cache locality by allocating raw memory
   */
  arrayedTree =  ArrayMalloc(Bucket*, (1 << 13) - 1); // TODO realloc
  arrayedTree[0] = new Bucket(defaultValue);
  minHeight = 0;
}
Hashly::~Hashly() {
  uint32_t upper = (1 << (minHeight + 1)) - 1;
  uint32_t lower = (1 << minHeight) - 1;
  for (uint32_t i = lower; i < upper; i++) {
    _free(i);
  }
  free(arrayedTree);
}

/**
 * Recursively free the buckets in arrayedTree
 */
void Hashly::_free(uint32_t i) {
  if (arrayedTree[i] == NULL) {
    _free((i << 1) + 1);
    _free((i << 1) + 2);
  } else {
    delete arrayedTree[i];
  }
}

void Hashly::_updateMinHeight(bool decrement) {
  minHeight -= decrement;
  uint32_t lower = (1 << minHeight) - 1;
  uint32_t upper = (lower << 1) | 1;
  bool res = true;
  for (uint32_t i = lower; i < upper && res; i++) {
    res = (arrayedTree[i] == NULL);
  }
  minHeight += res;
}

double Hashly::get(const char *key, int len) {
  Hash;
  I;
  HashlyFor {
    LocalBucket;
    IfBucket {
       return bucket->find(hash);
    }
    NextI;
  }
  return _defaultValue;
}

bool Hashly::set(const char *key, int len, double val) {
  Hash;
  I;
  HashlyFor {
    LocalBucket;
    IfBucket {
      uint8_t count = bucket->count;
      if (count < BUCKET_SIZE) {
        bucket->insert(key, val, hash);
        return true;
      } else { // it is full, so add a bucket
        arrayedTree[i] = NULL;
        uint32_t i_shift = i << 1;
        Bucket* leftChild = arrayedTree[i_shift | 1] = bucket;
        Bucket* rightChild = arrayedTree[i_shift + 2] = new Bucket(_defaultValue);
        uint32_t mask = 0x80000000 >> bit;
        for (uint8_t j = count; j--;) {
          Item item = bucket->items[j];
          if (item.hash & mask) {
            rightChild->insert(&item);
            bucket->items[j] = bucket->items[--count];
          }
        }
        leftChild->count = count;
        _updateMinHeight(false);
      }
    }
    NextI;
  }
  return false; // you never reach here, fake IDE
}

bool Hashly::has(const char *key, int len) {
  return get(key, len) != _defaultValue;
};

bool Hashly::del(const char *key, int len) {
  Hash;
  I;
  HashlyFor {
    LocalBucket;
    IfBucket {
      bool res = bucket->remove(hash);
      if (res && ! (bucket->count) && i) { // if it becomes empty after deletion
        uint32_t left = i & 1; // left is always an odd number
        uint32_t i_1 = i - 1;
        Bucket* sibling = arrayedTree[i_1 + (left << 1)];
        uint32_t parent = i_1 >> 1;
        if (sibling != NULL) { // if sibling does not have children
          arrayedTree[parent] = sibling;
          delete bucket;
          _updateMinHeight(true);
        }
      }
      return res;
    }
    NextI;
  }
  return false; // you never reach here, fake IDE
}